### PR TITLE
Initial attempt at a spec.

### PR DESCRIPTION
**Pull request by hober**

### CODE REVIEW

**Comment by othermaciej** [comment]
Looks like a good start!

File: storage-access.bs [hunk 1]
```diff
@@ -4,27 +4,111 @@ Shortname: storage-access
 Repository: privacycg/storage-access
 URL: https://privacycg.github.io/storage-access/
 Editor: John Wilander, w3cid 89478, Apple Inc. https://apple.com/, wilander@apple.com
-Abstract: TODO
-Status Text: This specification is intended to be merged into the HTML Living Standard. It is neither a WHATWG Living Standard nor a W3C standard.
+Abstract: The Storage Access API enables content in iframes to request access to cookies.
```
Comment (johnwilander) [inline storage-access.bs:7]: WebKit's implementation only covers cookies but I believe Mozilla's covers other means of storage. Maybe go with "website data such as cookies"?

File: storage-access.bs [hunk 2]
```diff
@@ -4,27 +4,111 @@ Shortname: storage-access
...
+When invoked, |document|.<dfn export method for=Document><code>requestStorageAccess()</code></dfn> must run these steps:
+
+1. Let |p| be [=a new promise=].
+1. If |document|'s [=has storage access flag=] is set, [=resolve=] |p| and abort these steps.
+1. If the |document|'s [=active sandboxing flag set=] has its [=sandboxed origin browsing context flag=] set, [=reject=] |p| and abort these steps.
+1. If |document|'s [=has storage access flag=] is unset, [=reject=] |p| and abort these steps.
+1. If |document|'s <a for=Document>browsing context</a> is a [=top-level browsing context=], [=resolve=] |p| and abort these steps.
+1. If |document|'s <a for=Document>browsing context</a> is [=same origin=] with its [=top-level browsing context=], [=resolve=] |p| and abort these steps.
+1. If the |document|'s {{Window}} object has [=transient activation=] and the |document|'s [=active sandboxing flag set=] has its [=sandbox storage access by user activation flag=] set, [=reject=] |p| and abort these steps.
+1. If |document|'s <a for=Document>browsing context</a>'s [=opener browsing context=] is not its [=top-level browsing context=], [=reject=] |p| and abort these steps.
```
Comment (johnwilander) [inline storage-access.bs:81]: I believe we have agreed on allowing nested iframes to request storage access so this step will likely be removed. I see you have a reference to that issue further down.

File: storage-access.bs [hunk 3]
```diff
@@ -4,27 +4,111 @@ Shortname: storage-access
...
+1. If the algorithm is invoked when the |document|'s {{Window}} object does not have [=transient activation=], [=reject=] |p| and abort these steps.
+1. <span class=XXX>Finish this algorithm.</span>
```
Comment (johnwilander) [inline storage-access.bs:83]: Something to consider is to add preserving the user gesture in all cases but explicit "Don't allow" from the user. That's what WebKit does to allow the iframe to do a popup if it turns out that the user is not logged in yet.

File: storage-access.bs [hunk 4]
```diff
@@ -4,27 +4,111 @@ Shortname: storage-access
...
+
+To the [=parse a sandboxing directive=] algorithm, add the following under step 3:
+
+<ul>
+<li>The [=sandbox storage access by user activation flag=], unless <var ignore>tokens</var> contains the <dfn export attr-value for=iframe/sandbox>allow-storage-access-by-user-activation</dfn> keyword.
+</ul>
+
+ISSUE(10): Provide mechanism for nested iframes to request storage access
+
+ISSUE(12): What about Feature Policy?
```
Comment (johnwilander) [inline storage-access.bs:111]: Will the change to the HTML Living Standard cover when storage access is removed, such as when the page is navigated? Or will it only cover how storage access is checked (hasStorageAccess) and requested (requestStorageAccess)?

Comment (hober) [reply storage-access.bs:118]: I've tried to capture this in d62d30c. Let me know if that works.

File: storage-access.bs (hunk 2 above)
Comment (hober) [reply storage-access.bs:84]: I think this is #10? If so, I'll drop this text when that issue gets resolved.

File: storage-access.bs (hunk 1 above)
Comment (hober) [reply storage-access.bs]: Fixed in 4ccd693.

File: storage-access.bs (hunk 3 above)
Comment (hober) [reply storage-access.bs]: I think I've done this in 8c9b716.

**Comment by hober** [comment]
Though there are a number of things left to spec, I think this is ready to go as a first draft. Please take a look, @johnwilander & @ehsan!

**Comment by johnwilander** [comment]
Looks good to me!

**Comment by hober** [comment]
Okay, great. My understanding is that @ehsan is on leave. @TanviHacks is going to drum up a review from someone familiar with the Gecko implementation, and assuming they think this looks okay, let's land it.

**Comment by annevk** [comment]
FWIW, others from Mozilla will be able to review this over the next couple of weeks, but that shouldn't block this from merging either. Issues can/will be filed as needed. Hope that helps.

**Comment by hober** [comment]
Thanks, @annevk! Changes based on your review (a couple of fixes and a lot of new issues inline) are in 507b672 & c94752f.

File: storage-access.bs [hunk 5]
```diff
@@ -4,27 +4,217 @@ Shortname: storage-access
...
+1. If |user expression of permission| is true, [=resolve=] |p|. <!-- WebKit's DocumentStorageAccess.cpp#L191 -->
+1. If |user expression of permission| is false, let |w| be |doc|'s {{Window}} object and run these steps:
+    1. If |w| has [=transient activation=] and |user expression of permission| is false, [=consume user activation=] with |w|. <!-- WebKit's DocumentStorageAccess.cpp#L181 -->
+    1. Unset |doc|'s [=has storage access flag=].
+    1. Set |doc|'s [=was expressly denied storage access flag=].
+    1. [=Reject=] |p|. <!-- WebKit's DocumentStorageAccess.cpp#L194 --> <!-- Gecko's Document.cpp#l15805 -->
+
+<h3 id="navigation">Changes to navigation</h3>
+
+Before changing the [=current entry=] of a [=session history=], unset the [=has storage access flag=] of the old [=current entry=]'s {{Document}}, if it has one.
```
[the remark this answers is not in the document]
Comment (hober) [reply storage-access.bs:160]: Indeed. I've marked it as such in an inline issue.

File: storage-access.bs [hunk 6]
```diff
@@ -4,27 +4,217 @@ Shortname: storage-access
...
+1. Assert: |topDoc| is the [=active document=] of |doc|'s <a for=Document>browsing context</a>'s [=top-level browsing context=].
+1. If |doc|'s [=has storage access flag=] is set, return true.
+1. Let |has storage access| (a {{boolean}}) be the result of running a UA-defined set of steps to determine if |doc| has storage access when it is loaded in a third-party context on |topDoc|.
+1. If |has storage access| is true, set |doc|'s [=has storage access flag=].
+1. Return |has storage access|.
+
+When required to <dfn type="abstract-op">determine the storage access policy</dfn> for {{Document|Documents}} |doc| and |topDoc| with {{Promise}} |p|, run these steps:
+
+1. Assert: |topDoc| is the [=active document=] of |doc|'s <a for=Document>browsing context</a>'s [=top-level browsing context=].
+1. Let |should implicitly grant| and |should implicitly deny| (both {{boolean|booleans}}) be the result of running a UA-defined set of steps to determine if |doc|'s request for storage access on |topDoc| should be granted or denied without prompting the user.
```
[the remark this answers is not in the document]
Comment (hober) [reply storage-access.bs]: Name fixed. Second point marked as an inline issue.

File: storage-access.bs [hunk 7]
```diff
@@ -4,27 +4,217 @@ Shortname: storage-access
...
+1. If the algorithm is invoked when |doc|'s {{Window}} object does not have [=transient activation=], [=reject=] |p| and abort these steps. <!-- WebKit's DocumentStorageAccess.cpp#L163 --> <!-- Gecko's Document.cpp#l15680 -->
+1. [=Determine the storage access policy=] with |doc|, |topDoc|, and |p| and abort these steps. <!-- WebKit's DocumentStorageAccess.cpp#L177 --> <!-- Gecko's Document.cpp#l15685 -->
+1. Set |doc|'s [=has storage access flag=], [=resolve=] |p|, and abort these steps. <!-- Gecko's Document.cpp#l15805 -->
+1. [=If aborted=], return |p|.
+
+ISSUE(10): Remove step 9 if we determine that nested <{iframe}>s should be able to request storage access.
+
+<h4 id="ua-policy">User Agent storage access policies</h4>
+
+Different User Agents have different policies around whether or not third-party <{iframe}>'s may access data placed into client-side storage mechanisms when the <{iframe}>'s {{Document}}'s <a for=Document>origin</a> was loaded in a first-party context. User Agents check these policies when client-side storage is accessed (see [[#storage]]) as well as by {{Document/hasStorageAccess()}} and {{Document/requestStorageAccess()}}.
```
[the remark this answers is not in the document]
Comment (hober) [reply storage-access.bs]: Yeah, I've made a note to simply remove the use of both terms throughout the document.

File: storage-access.bs [hunk 8]
```diff
@@ -4,27 +4,217 @@ Shortname: storage-access
...
+1. If |doc| is [=same origin=] with |topDoc|, [=resolve=] |p| and abort these steps. <!-- WebKit's DocumentStorageAccess.cpp#L146 --> <!-- Gecko's Document.cpp#l15604 --> <!-- Gecko's Document.cpp#l15657 -->
+1. If |doc|'s [=active sandboxing flag set=] has its [=sandbox storage access by user activation flag=] set, [=reject=] |p| and abort these steps. <!-- WebKit's DocumentStorageAccess.cpp#L152 --> <!-- Gecko's Document.cpp#l15667 -->
+1. If |doc|'s <a for=Document>browsing context</a>'s [=opener browsing context=] is not its [=top-level browsing context=], [=reject=] |p| and abort these steps. <!-- WebKit's DocumentStorageAccess.cpp#L158 --> <!-- Gecko's Document.cpp#l15673 -->
+1. If the algorithm is invoked when |doc|'s {{Window}} object does not have [=transient activation=], [=reject=] |p| and abort these steps. <!-- WebKit's DocumentStorageAccess.cpp#L163 --> <!-- Gecko's Document.cpp#l15680 -->
+1. [=Determine the storage access policy=] with |doc|, |topDoc|, and |p| and abort these steps. <!-- WebKit's DocumentStorageAccess.cpp#L177 --> <!-- Gecko's Document.cpp#l15685 -->
```
[the remark this answers is not in the document]
Comment (hober) [reply storage-access.bs]: Good point. I've marked this inline.

File: storage-access.bs [hunk 9]
```diff
@@ -4,27 +4,217 @@ Shortname: storage-access
...
+
+When invoked on {{Document}} |doc|, the <dfn export method for=Document><code>hasStorageAccess()</code></dfn> method must run these steps:
+
+<!-- https://developer.mozilla.org/en-US/docs/Web/API/Document/hasStorageAccess -->
+<!-- https://trac.webkit.org/browser/webkit/trunk/Source/WebCore/dom/DocumentStorageAccess.cpp#L80 -->
+<!-- https://hg.mozilla.org/mozilla-central/file/tip/dom/base/Document.cpp#l15512 -->
+
+1. Let |p| be [=a new promise=].
+1. If |doc|'s [=was expressly denied storage access flag=] is set, [=resolve=] |p| with false and abort these steps. <!-- WebKit's DocumentStorageAccess.cpp#L85 -->
+1. If |doc|'s [=active sandboxing flag set=] has its [=sandboxed origin browsing context flag=] set, [=resolve=] |p| with false and abort these steps. <!-- WebKit's DocumentStorageAccess.cpp#L90 --> <!-- Gecko's Document.cpp#l15526 -->
```
[the remark this answers is not in the document]
Comment (hober) [reply storage-access.bs]: Marked inline.

File: storage-access.bs [hunk 10]
```diff
@@ -4,27 +4,217 @@ Shortname: storage-access
...
+1. If |doc|'s <a for=Document>browsing context</a> is a [=top-level browsing context=], [=resolve=] |p| with true and abort these steps. <!-- WebKit's DocumentStorageAccess.cpp#L95 --> <!-- Gecko's Document.cpp#l15531 -->
+1. Let |topDoc| be the [=active document=] of |doc|'s <a for=Document>browsing context</a>'s [=top-level browsing context=].
+1. If |doc| is [=same origin=] with |topDoc|, [=resolve=] |p| with true and abort these steps. <!-- WebKit's DocumentStorageAccess.cpp#L102 --> <!-- Gecko's Document.cpp#l15541 -->
+1. Resolve |p| with the result of running [=determine if a document has storage access=] with |doc| and |topDoc| and abort these steps. <!-- WebKit's DocumentStorageAccess.cpp#L115 --> <!-- Gecko's Document.cpp#l15548 -->
+1. [=If aborted=], return |p|.
```
[the remark this answers is not in the document]
Comment (hober) [reply storage-access.bs]: Fixed.

File: storage-access.bs [hunk 11]
```diff
@@ -4,27 +4,217 @@ Shortname: storage-access
...
+
+<!-- https://developer.mozilla.org/en-US/docs/Web/API/Document/hasStorageAccess -->
+<!-- https://trac.webkit.org/browser/webkit/trunk/Source/WebCore/dom/DocumentStorageAccess.cpp#L80 -->
+<!-- https://hg.mozilla.org/mozilla-central/file/tip/dom/base/Document.cpp#l15512 -->
+
+1. Let |p| be [=a new promise=].
+1. If |doc|'s [=was expressly denied storage access flag=] is set, [=resolve=] |p| with false and abort these steps. <!-- WebKit's DocumentStorageAccess.cpp#L85 -->
+1. If |doc|'s [=active sandboxing flag set=] has its [=sandboxed origin browsing context flag=] set, [=resolve=] |p| with false and abort these steps. <!-- WebKit's DocumentStorageAccess.cpp#L90 --> <!-- Gecko's Document.cpp#l15526 -->
+1. If |doc|'s <a for=Document>browsing context</a> is a [=top-level browsing context=], [=resolve=] |p| with true and abort these steps. <!-- WebKit's DocumentStorageAccess.cpp#L95 --> <!-- Gecko's Document.cpp#l15531 -->
+1. Let |topDoc| be the [=active document=] of |doc|'s <a for=Document>browsing context</a>'s [=top-level browsing context=].
```
[the remark this answers is not in the document]
Comment (hober) [reply storage-access.bs:100]: Noted inline.

File: storage-access.bs [hunk 12]
```diff
@@ -4,27 +4,217 @@ Shortname: storage-access
...
+Each {{Document}} has an associated <dfn export for=Document id=was-expressly-denied-storage-access-flag>was expressly denied storage access flag</dfn>, initially unset.
+
+When invoked on {{Document}} |doc|, the <dfn export method for=Document><code>hasStorageAccess()</code></dfn> method must run these steps:
+
+<!-- https://developer.mozilla.org/en-US/docs/Web/API/Document/hasStorageAccess -->
+<!-- https://trac.webkit.org/browser/webkit/trunk/Source/WebCore/dom/DocumentStorageAccess.cpp#L80 -->
+<!-- https://hg.mozilla.org/mozilla-central/file/tip/dom/base/Document.cpp#l15512 -->
+
+1. Let |p| be [=a new promise=].
+1. If |doc|'s [=was expressly denied storage access flag=] is set, [=resolve=] |p| with false and abort these steps. <!-- WebKit's DocumentStorageAccess.cpp#L85 -->
```
[the remark this answers is not in the document]
Comment (hober) [reply storage-access.bs]: Fixed.

File: storage-access.bs [hunk 13]
```diff
@@ -4,27 +4,217 @@ Shortname: storage-access
...
+}
+</style>
+
 <section class="non-normative">
 <h2 id="intro">Introduction</h2>
 
 <em>This section is non-normative.</em>
 
-TODO
+User Agents sometimes block access to client-side storage mechanisms in third-party contexts. This can break authenticated embeds such as commenting widgets, which often rely on cookies for authentication.
```
[the remark this answers is not in the document]
Comment (hober) [reply storage-access.bs]: I've made a note to fix this.

File: storage-access.bs [hunk 14]
```diff
@@ -4,27 +4,217 @@ Shortname: storage-access
...
+
+The Storage Access API enables cross-origin <{iframe}>s to request and be granted access to their client-side storage, so that authenticated embeds can work in such User Agents. [[STORAGE-ACCESS-INTRO]]
```
[the remark this answers is not in the document]
Comment (hober) [reply storage-access.bs]: Good point. Noted inline.

**Comment by Brandr0id** [comment]
@hober thanks! Looks good as a first draft.

**Pull request merged**